trie/iterator.go

```diff
@@ -535,6 +535,44 @@ func (it *nodeIterator) pop() {
 	it.stack = it.stack[:len(it.stack)-1]
 }
 
+// reachedPath normalizes a path by truncating a terminator if present, and returns true if it is
+// greater than or equal to the target.
+func reachedPath(path, target []byte) bool {
+	if hasTerm(path) {
+		path = path[:len(path)-1]
+	}
+	return bytes.Compare(path, target) >= 0
+}
+
+// A value embedded in a fullNode occupies its last child slot, but we want to visit it last to
+// produce a pre-order traversal. These two functions handle the ordering by mapping index 0 to the
+// value, and shifting the indices of children over by 1.
+func prevChildIndex(index int) int {
+	switch index {
+	case 0:
+		return 16
+	case 16:
+		return -1
+	case 17:
+		return 15
+	default:
+		return index - 1
+	}
+}
+
+func nextChildIndex(index int) int {
+	switch index {
+	case -1:
+		return 16
+	case 15:
+		return 17
+	case 16:
+		return 0
+	default:
+		return index + 1
+	}
+}
```

Comment:
I don't understand why the `next` for `0` is `16`. Nor why `15` leads to `17`, which then continues indefinitely. Please add some comments

Reply:
> In order to properly change the order of traversal, I added a fix that will jump to the full node's leaf first, then back to iterate its child nodes normally, and skip the leaf at the end. 

Sorry, my bad for not reading properly. Still, it's a bit funky behaviour... Would definitely need more comments in the code

Reply:
True, I will add some clarification.

Reply:
I've expanded on the comments, hope it's clearer.